queries/sway/highlights.scm

```diff
@@ -0,0 +1,336 @@
+
+(shorthand_field_identifier) @variable.member
+
+(loop_label
+  "'" @label
+  (identifier) @label)
+
+; ---
+; Punctuation
+; ---
+[
+  "::"
+  ":"
+  "."
+  ";"
+  ","
+] @punctuation.delimiter
+
+[
+  "("
+  ")"
+  "["
+  "]"
+  "{"
+  "}"
+  "#"
+] @punctuation.bracket
+
+(type_arguments
+  [
+    "<"
+    ">"
+  ] @punctuation.bracket)
+
+(type_parameters
+  [
+    "<"
+    ">"
+  ] @punctuation.bracket)
+
+(closure_parameters
+  "|" @punctuation.bracket)
+
+(let_declaration
+  pattern: [
+    (identifier) @variable
+    (tuple_pattern
+      (identifier) @variable)
+  ])
+
+; It needs to be anonymous to not conflict with `call_expression` further below.
+(_
+  value: (field_expression
+    value: (identifier)? @variable
+    field: (field_identifier) @variable.member))
+
+(parameter
+  pattern: (identifier) @variable.parameter)
+
+(parameter
+  pattern: (ref_pattern
+    [
+      (mut_pattern
+        (identifier) @variable.parameter)
+      (identifier) @variable.parameter
+    ]))
+
+(closure_parameters
+  (identifier) @variable.parameter)
+
+(for_expression
+  "for" @keyword.repeat)
+
+"in" @keyword.repeat
+
+[
+  "match"
+  "if"
+  "else"
+] @keyword.conditional
+
+"while" @keyword.repeat
+
+[
+  "break"
+  "continue"
+  "return"
+  "yield"
+] @keyword.return
+
+"use" @keyword.import
+
+(mod_item
+  "mod" @keyword.import
+  !body)
+
+(use_as_clause
+  "as" @keyword.import)
+
+(type_cast_expression
+  "as" @keyword.operator)
+
+[
+  "as"
+  "mod"
+  "abi"
+  "impl"
+  "where"
+  "trait"
+  "for"
+  "let"
+  "contract"
+  "script"
+  "predicate"
+  "library"
+] @keyword
+
+[
+  "struct"
+  "enum"
+  "storage"
+  "configurable"
+  "type"
+] @keyword.type
+
+[
+  "fn"
+  "abi"
+] @keyword.function
+
+[
+  (mutable_specifier)
+  "const"
+  "ref"
+  "deref"
+  "move"
+  "pub"
+] @keyword.modifier
+
+(reference_type
+  "&" @keyword.modifier)
+
+(self_parameter
+  "&" @keyword.modifier)
+
+; -------
+; Guess Other Types
+; -------
+((identifier) @constant
+  (#match? @constant "^[A-Z][A-Z\\d_]*$"))
+
+; ---
+; PascalCase identifiers in call_expressions (e.g. `Ok()`)
+; are assumed to be enum constructors.
+; ---
+(call_expression
+  function: [
+    ((identifier) @type.enum.variant
+      (#match? @type.enum.variant "^[A-Z]"))
+    (scoped_identifier
+      name: ((identifier) @type.enum.variant
+        (#match? @type.enum.variant "^[A-Z]")))
+  ])
+
+; ---
+; Assume that types in match arms are enums and not
+; tuple structs. Same for `if let` expressions.
+; ---
+(match_pattern
+  (scoped_identifier
+    name: (identifier) @constructor))
+
+(tuple_struct_pattern
+  type: [
+    (identifier) @constructor
+    (scoped_identifier
+      name: (identifier) @constructor)
+  ])
+
+(struct_pattern
+  type: [
+    (type_identifier) @constructor
+    (scoped_type_identifier
+      name: (type_identifier) @constructor)
+  ])
+
+; ---
+; Other PascalCase identifiers are assumed to be structs.
+; ---
+((identifier) @type
+  (#match? @type "^[A-Z]"))
```

Comment:
You should know better ;)

Reply:
Yeah, I was wondering about whether to remove though because e.g. rust has many such captures? I should mention this language is nearly identical to rust (but doesn't inherit it annoyingly). Seems weird to arbitrarily disallow these kinds of queries while many languages still have them

Reply:
I was referring to the `#match?` itself...

(I agree that we should remove the "capital convention" everywhere or nowhere -- with a personal preference for "everywhere"; the point of tree-sitter is that we can do better than that kind of drunk aiming.)

Reply:
Oh oops :laughing: thank you